File: 494-target-sum/target-sum.py
class Solution:
    def findTargetSumWays(self, nums: List[int], target: int) -> int:
        

        # Approach -
        # 1. 1 + 1 + 1 + 1 - 1 => 3
        # P - N = target
        # P + N = totalSum
        # 2P = target + totalSum
        # P = (target + totalSum)/2
        dp = {}

        def generate(index, currentSum):
            if (index, currentSum) in dp:
                return dp[(index, currentSum)]

            if index == len(nums):
                if currentSum == target:
                    return 1
                else:
                    return 0

            count_positive = generate(index + 1, currentSum + nums[index])
            count_negative = generate(index + 1, currentSum - nums[index])

            dp[(index, currentSum)] = count_negative + count_positive

            return dp[(index, currentSum)]

        return generate(0, 0)

File: 494-target-sum/target-sum.java
class Solution {
    public int findTargetSumWays(int[] nums, int target) {
        Map<String, Integer> memo = new HashMap<>();
        return dfs(nums, 0, 0, target, memo);
    }

    private int dfs(int[] nums, int index, int currentSum, int target, Map<String, Integer> memo) {
        String key = index + "," + currentSum;

        if (memo.containsKey(key)) {
            return memo.get(key);
        }

        if (index == nums.length) {
            return currentSum == target ? 1 : 0;
        }

        int positive = dfs(nums, index + 1, currentSum + nums[index], target, memo);
        int negative = dfs(nums, index + 1, currentSum - nums[index], target, memo);

        memo.put(key, positive + negative);
        return memo.get(key);
    }
}